### DISCUSSION TITLE
Is there a recommended way to play local audio files in a WASM app?

### DISCUSSION THREAD

**Discussion by BrianDT** [unoplatform/uno#12463] [Q&A]
The Plugin.Maui.Audio does a good job on Windows, Android and iOS but there is nothing for WASM.

Is there a recommended solution for WASM apps? Something with a similar interface to IAudioPlayer from the plugin would be nice.

**Comment by jeromelaban** [comment]
Thanks for the question. We introduced [MediaPlayerElement](https://platform.uno/docs/articles/controls/MediaPlayerElement.html?q=mediap) for WebAssembly, you may be able to play content this way.

**Reply by BrianDT** [reply]
@jeromelaban  I noted this in the 4.9 release announcement.

I have created a simple sample based on the latest templates and while the MediaPlayerElement works well on Windows I have not been able to play any sounds either interactively or programmatically on WASM, Android or iOS.

I have uploaded the sample here:
https://github.com/BrianDT/UnoProgMediaPlayerElementIssue

**Comment by mrlacey** [comment]
This looks to be related to the more general question at https://github.com/unoplatform/uno/discussions/10129

**Reply by BrianDT** [reply]
@mrlacey Yes for the background audio part, however my sample is not playing audio mp3 files interactively either.

**Reply by jeromelaban** [reply]
You should be able to play audio files properly, see https://playground.platform.uno/#30ab5832. This particular sample is hitting https://github.com/unoplatform/uno/issues/12533 though, which is a browser playback restriction. Is this what you're seeing?

**Comment by BrianDT** [comment]
@jeromelaban Looking at the sample you referenced, I can see that this accesses a public web based audio file while I want to access a local project based audio file, as in the example Source="ms-appx:///CommonMedia/bing.mp3"

I have updated the example to include the option of playing a remote media file.

This is a summary of my findings:

| Platform  | Local file | Remote file Auto play true | Remote file Auto play false |
| --- | --- |--- | --- |
| Windows | OK | OK | OK | OK |
| WASM | Player displays, Audio does not start when play pressed |- Player displays, Audio does not start but can be paused and restarted and can be stopped KNOWN ISSUE | Displays without transport controls cannot be started |
| iOS | Player displays, Audio does not start when play pressed |- Black rectangle on screen Plays Audio which cannot be stopped | Black rectangle on screen Auto cannot be started |
| Android | Player displays, Audio does not start when play pressed |- Player displays, Audio starts and can be stopped | Displays without transport controls cannot be started |

Also, I note that on iOS and Android AutoPlay defaults to true if unset whereas it should default to false.

**Reply by jeromelaban** [reply]
Thanks for the update.

Local sources are not yet supported (https://github.com/unoplatform/uno/issues/12331), the transport controls not appearing with autoplay false is reported in https://github.com/unoplatform/uno/issues/12535.

Could you provide the XAML for the "Remote file Auto play true" ?

We're tracking all the MediaPlayer issues here: https://github.com/unoplatform/uno/issues/2642

**Comment by BrianDT** [comment]
@jeromelaban  The XAML is identical to the playground sample you referenced

'''
        <ios:TextBlock Text="Web media file with AutoPlay set to True" HorizontalAlignment="Center" Margin="0,24,0,8" />
        <ios:MediaPlayerElement x:Name="mediaPlayerRemoteMediaAuto"
                            AreTransportControlsEnabled="True"
                            Source="https://uno-assets.platform.uno/tests/audio/Getting_Started_with_Uno_Platform_and_Visual_Studio_Code.mp3"
                            AutoPlay="True"/>

'''
This is the result on iOS
![IMG_0002](https://github.com/unoplatform/uno/assets/4189260/4cfcc290-86ff-45b3-944f-0a3da6494497)

**Reply by jeromelaban** [reply]
Thanks. This is being addressed in https://github.com/unoplatform/uno/pull/12421

**Comment by DylanSafeAt** [comment]
@jeromelaban Does the Media Player Element support videos with urls that do not end in .mp4? When I set the source of the element to link to a Vimeo video, nothing loads, however if I set the source to a video with a url ending in .mp4 it works fine.

Thanks for any help!

**Reply by jeromelaban** [reply]
Thanks for the question. Is it for Wasm, specifically? If it is, it may be a restriction of the `media` element in HTML. If you open the browser debugger, you may see additional information about the loading of the content.

**Reply by DylanSafeAt** [reply]
Thanks for the quick response! Yes it is for WASM specifically, I will try that out. Thanks again.

**Reply by DylanSafeAt** [reply]
@jeromelaban would you know where to find that in the inspector? I looking through the network tab and nothing shows unless I use a .mp4 link. Makes me think that its something with the code.

**Reply by jeromelaban** [reply]
It looks like we're only supporting this formats: https://github.com/unoplatform/uno/blob/3ddcef8fb75a6da6f86a8abf06db004240bbacbf/src/AddIns/Uno.UI.MediaPlayer.WebAssembly/HtmlMediaPlayer.cs#L27. If you're pasting the URL to your video directly in the adress bar, does it play?
